File: crates/substrate-core/src/space.rs
use std::fs;
use std::path::{Path, PathBuf};
use std::sync::atomic::{AtomicU64, Ordering};

use serde::{Deserialize, Serialize};

use crate::error::{Result, SubstrateError};
use crate::name::Name;

/// Everything substrate writes for a space lives under `.substrate/` in the
/// project root — like `.git/`, one hidden directory, project untouched.
pub const SUBSTRATE_DIR: &str = ".substrate";
pub const SPACE_CONFIG_FILE: &str = ".substrate/config.yaml";
pub const THREADS_DIR: &str = ".substrate/threads";
pub const THREAD_CONFIG_FILE: &str = "config.yaml";

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SpaceConfig {
    #[serde(default = "default_version")]
    pub version: u32,
    #[serde(default)]
    pub participants: Vec<Participant>,
}

fn default_version() -> u32 {
    1
}

impl Default for SpaceConfig {
    fn default() -> Self {
        SpaceConfig {
            version: 1,
            participants: Vec::new(),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Participant {
    pub name: Name,
    pub kind: ParticipantKind,
}

/// Descriptive metadata only (display, listings). The turn engine never
/// branches on kind: humans, agents, and anything else are peers in the room.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum ParticipantKind {
    Human,
    Agent,
    Other,
}

impl std::fmt::Display for ParticipantKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(match self {
            ParticipantKind::Human => "human",
            ParticipantKind::Agent => "agent",
            ParticipantKind::Other => "other",
        })
    }
}

impl std::str::FromStr for ParticipantKind {
    type Err = String;

    fn from_str(s: &str) -> std::result::Result<Self, String> {
        match s {
            "human" => Ok(ParticipantKind::Human),
            "agent" => Ok(ParticipantKind::Agent),
            "other" => Ok(ParticipantKind::Other),
            _ => Err(format!("unknown kind '{s}' (human|agent|other)")),
        }
    }
}

/// Handle to a space root directory. Every operation re-reads from disk:
/// the filesystem is the shared state for all processes in the room, so
/// nothing is cached.
#[derive(Debug, Clone)]
pub struct Space {
    root: PathBuf,
}

impl Space {
    /// Create a new space: the directory (if needed) plus `.substrate/`.
    pub fn init(root: &Path) -> Result<Space> {
        fs::create_dir_all(root.join(THREADS_DIR))?;
        let config_path = root.join(SPACE_CONFIG_FILE);
        if config_path.exists() {
            return Err(SubstrateError::Io(std::io::Error::new(
                std::io::ErrorKind::AlreadyExists,
                format!("{} already exists", config_path.display()),
            )));
        }
        let space = Space {
            root: root.to_path_buf(),
        };
        space.save_config(&SpaceConfig::default())?;
        Ok(space)
    }

    pub fn open(root: &Path) -> Result<Space> {
        if !root.join(SPACE_CONFIG_FILE).is_file() {
            return Err(SubstrateError::NotASpace(root.to_path_buf()));
        }
        Ok(Space {
            root: root.to_path_buf(),
        })
    }

    pub fn root(&self) -> &Path {
        &self.root
    }

    pub fn config(&self) -> Result<SpaceConfig> {
        let content = fs::read_to_string(self.root.join(SPACE_CONFIG_FILE))?;
        Ok(serde_norway::from_str(&content)?)
    }

    pub fn save_config(&self, config: &SpaceConfig) -> Result<()> {
        let yaml = serde_norway::to_string(config)?;
        write_atomic(&self.root.join(SPACE_CONFIG_FILE), &yaml)
    }

    pub fn add_participant(&self, name: Name, kind: ParticipantKind) -> Result<()> {
        let mut config = self.config()?;
        if config.participants.iter().any(|p| p.name == name) {
            return Err(SubstrateError::DuplicateParticipant(name));
        }
        config.participants.push(Participant { name, kind });
        self.save_config(&config)
    }

    pub fn participant(&self, name: &str) -> Result<Participant> {
        self.config()?
            .participants
            .into_iter()
            .find(|p| p.name.as_str() == name)
            .ok_or_else(|| SubstrateError::UnknownParticipant(name.to_string()))
    }

    /// The space's `.substrate/` directory — the only place substrate
    /// writes, and the right thing to file-watch.
    pub fn substrate_dir(&self) -> PathBuf {
        self.root.join(SUBSTRATE_DIR)
    }

    pub fn thread_dir(&self, thread: &Name) -> PathBuf {
        self.root.join(THREADS_DIR).join(thread.to_path_component())
    }

    /// All threads in the space: directories under `.substrate/threads/`
    /// containing a config.yaml.
    pub fn list_threads(&self) -> Result<Vec<Name>> {
        let mut found = Vec::new();
        let threads_root = self.root.join(THREADS_DIR);
        if !threads_root.is_dir() {
            return Ok(found); // hand-made or empty space: just no threads
        }
        for dir_entry in fs::read_dir(&threads_root)? {
            let dir_entry = dir_entry?;
            if !dir_entry.path().join(THREAD_CONFIG_FILE).is_file() {
                continue;
            }
            if let Some(name) = dir_entry
                .file_name()
                .to_str()
                .and_then(|s| Name::from_path_component(s).ok())
            {
                found.push(name);
            }
        }
        found.sort();
        Ok(found)
    }
}

static TMP_COUNTER: AtomicU64 = AtomicU64::new(0);

/// Write via dotfile temp + rename (atomic on APFS and other POSIX
/// filesystems). Readers skip non-entry filenames, so a torn write is
/// never visible to anyone in the room.
pub(crate) fn write_atomic(path: &Path, content: &str) -> Result<()> {
    let dir = path.parent().unwrap_or_else(|| Path::new("."));
    let tmp = dir.join(format!(
        ".tmp-{}-{}",
        std::process::id(),
        TMP_COUNTER.fetch_add(1, Ordering::Relaxed)
    ));
    fs::write(&tmp, content)?;
    fs::rename(&tmp, path)?;
    Ok(())
}

File: internal/substrate/space.go
// Package substrate implements the turn-based space engine and its on-disk
// format: spaces, participants, threads, entries, and the floor-passing rules
// that every front end (CLI, TUI, MCP server, HTTP proxy) shares.
package substrate

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"slices"

	"gopkg.in/yaml.v3"
)

const (
	SubstrateDir     = ".substrate"
	SpaceConfigFile  = ".substrate/config.yaml"
	ThreadsDir       = ".substrate/threads"
	ThreadConfigFile = "config.yaml"
)

// ParticipantKind classifies a participant: human, agent, or other.
type ParticipantKind string

const (
	Human ParticipantKind = "human"
	Agent ParticipantKind = "agent"
	Other ParticipantKind = "other"
)

func ParseParticipantKind(raw string) (ParticipantKind, error) {
	switch ParticipantKind(raw) {
	case Human, Agent, Other:
		return ParticipantKind(raw), nil
	default:
		return "", fmt.Errorf("unknown kind %q (human|agent|other)", raw)
	}
}

func (k *ParticipantKind) UnmarshalYAML(node *yaml.Node) error {
	if node.Kind != yaml.ScalarNode {
		return fmt.Errorf("participant kind must be a string")
	}
	parsed, err := ParseParticipantKind(node.Value)
	if err != nil {
		return err
	}
	*k = parsed
	return nil
}

func (k ParticipantKind) MarshalYAML() (any, error) { return string(k), nil }

type Participant struct {
	Name Name            `yaml:"name"`
	Kind ParticipantKind `yaml:"kind"`
}

type SpaceConfig struct {
	Version      int            `yaml:"version"`
	Participants []Participant  `yaml:"participants,omitempty"`
	Extra        map[string]any `yaml:",inline"`
}

// Space is a directory containing a .substrate tree. All state lives on disk;
// a Space holds only the root path and re-reads files on every operation.
type Space struct{ root string }

// InitSpace creates the .substrate tree under root; it fails if one exists.
func InitSpace(root string) (*Space, error) {
	root = filepath.Clean(root)
	configPath := filepath.Join(root, SpaceConfigFile)
	if _, err := os.Stat(configPath); err == nil {
		return nil, fmt.Errorf("%s already exists", configPath)
	} else if !errors.Is(err, os.ErrNotExist) {
		return nil, err
	}
	if err := ensureDir(filepath.Join(root, ThreadsDir)); err != nil {
		return nil, err
	}
	space := &Space{root: root}
	if err := space.SaveConfig(SpaceConfig{Version: 1}); err != nil {
		return nil, err
	}
	return space, nil
}

// OpenSpace opens an existing space rooted at root.
func OpenSpace(root string) (*Space, error) {
	root = filepath.Clean(root)
	info, err := os.Stat(filepath.Join(root, SpaceConfigFile))
	if err != nil || !info.Mode().IsRegular() {
		return nil, &NotASpaceError{Root: root}
	}
	return &Space{root: root}, nil
}

func (s *Space) Root() string         { return s.root }
func (s *Space) SubstrateDir() string { return filepath.Join(s.root, SubstrateDir) }
func (s *Space) ThreadDir(thread Name) string {
	return filepath.Join(s.root, ThreadsDir, thread.ToPathComponent())
}

func (s *Space) Config() (SpaceConfig, error) {
	data, err := os.ReadFile(filepath.Join(s.root, SpaceConfigFile))
	if err != nil {
		return SpaceConfig{}, err
	}
	var cfg SpaceConfig
	if err := yaml.Unmarshal(data, &cfg); err != nil {
		return SpaceConfig{}, err
	}
	if cfg.Version == 0 {
		cfg.Version = 1
	}
	return cfg, nil
}

func (s *Space) SaveConfig(cfg SpaceConfig) error {
	data, err := yaml.Marshal(cfg)
	if err != nil {
		return err
	}
	return writeAtomic(filepath.Join(s.root, SpaceConfigFile), data)
}

// AddParticipant registers a new participant in the space config.
func (s *Space) AddParticipant(name Name, kind ParticipantKind) error {
	return withFileLock(filepath.Join(s.SubstrateDir(), ".space.lock"), func() error {
		cfg, err := s.Config()
		if err != nil {
			return err
		}
		for _, p := range cfg.Participants {
			if p.Name == name {
				return &DuplicateParticipantError{Name: name}
			}
		}
		cfg.Participants = append(cfg.Participants, Participant{Name: name, Kind: kind})
		return s.SaveConfig(cfg)
	})
}

// Participant looks up a registered participant by name.
func (s *Space) Participant(name Name) (Participant, error) {
	cfg, err := s.Config()
	if err != nil {
		return Participant{}, err
	}
	for _, p := range cfg.Participants {
		if p.Name == name {
			return p, nil
		}
	}
	return Participant{}, &UnknownParticipantError{Name: name.String()}
}

// ListThreads returns the space's thread names in sorted order.
func (s *Space) ListThreads() ([]Name, error) {
	entries, err := os.ReadDir(filepath.Join(s.root, ThreadsDir))
	if errors.Is(err, os.ErrNotExist) {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	threads := make([]Name, 0, len(entries))
	for _, entry := range entries {
		if !entry.IsDir() {
			continue
		}
		if info, err := os.Stat(filepath.Join(s.root, ThreadsDir, entry.Name(), ThreadConfigFile)); err != nil || !info.Mode().IsRegular() {
			continue
		}
		name, err := NameFromPathComponent(entry.Name())
		if err == nil {
			threads = append(threads, name)
		}
	}
	slices.Sort(threads)
	return threads, nil
}

// EnsureParticipant registers name as an agent if it is not yet a participant.
// It reports whether a new registration happened.
func (s *Space) EnsureParticipant(name Name) (bool, error) {
	_, err := s.Participant(name)
	if err == nil {
		return false, nil
	}
	var unknown *UnknownParticipantError
	if !errors.As(err, &unknown) {
		return false, err
	}
	if err := s.AddParticipant(name, Agent); err != nil {
		return false, err
	}
	return true, nil
}
